### PR TITLE
LG/DCOS-49954: new section for ZooKeeper-specific recommendations

### PR DESCRIPTION
**Pull request by lsgunnlsgunn**
## Description

https://jira.mesosphere.com/browse/DCOS-49954
https://jira.mesosphere.com/browse/DCOS-50570
Tangentially:
https://jira.mesosphere.com/browse/COPS-4538
https://jira.mesosphere.com/browse/DCOS-21142

## Urgency
- [ ] Blocker 
- [x] High
- [ ] Medium

## Requirements
- Test all commands and procedures.
- Add [redirects](https://github.com/mesosphere/dcos-docs-site/wiki/Redirects).
- Change all affected versions (e.g. 1.7, 1.8, 1.9, 1.10, 1.11, 1.12).
- See the [contribution guidelines](https://github.com/mesosphere/dcos-docs-site/wiki/Contributing).

### CODE REVIEW

**Comment by gpaul** [comment]
Thanks, valuable info here

**Comment by lsgunnlsgunn** [comment]
Any comments or changes before I merge to 1.13 for GA?

**Comment by lsgunnlsgunn** [comment]
The suggestions look good to me. I'll let it bake for another day to give security team or others a chance to chime in. Worst case, I'll have to make additional changes in a new PR.

File: pages/1.13/installing/production/system-requirements/zookeeper-requirements/index.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+- ZooKeeper processes each read or write operation as a complete transaction rather than performing partial or incremental updates when there are changes to nodes in the cluster.
+- ZooKeeper does not remove old snapshots and log files from the file system. You must manage these files manually.
+
+ZooKeeper stores both the snapshots and transaction log in the location you specify using the `dataDir` parameter. You can also configure ZooKeeper to write the transaction log to a separate location using the `dataLogDir` parameter. Specifying both the `dataLogDir` and `dataDir` locations allows you to use separate file system directories and a dedicated log device. Using a dedicated log device for ZooKeeper transaction logs helps to avoid resource contention and latency issues.
+
+# Identifying potential problems
+One key way you can identify issues that are related to ZooKeeper is by searching ZooKeeper log files for messages related to synchronization (`fsync`) operations. For example, if the ZooKeeper data and log directories are mounted on the same disk or there are disk or network latency issues, the transaction log records messages similar to the following:
```
[the remark this answers is not in the document]
Comment (lsgunnlsgunn) [reply pages/1.13/installing/production/system-requirements/zookeeper-requirements/index.md]: Done.

File: pages/1.13/installing/production/system-requirements/zookeeper-requirements/index.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+Isolate master nodes from agent nodes to prevent memory and CPU contention, especially when workloads are started on the agents.
+
+The most common ZooKeeper issues occur when ZooKeeper transaction log files are stored on the same solid state drive (SSD) as other components where ZooKeeper write performance becomes a bottleneck that slows down or delays processing of other worker threads.
+    
+ZooKeeper should not share resources with any other processes or services. ZooKeeper writes new transactions to the log before performing updates or sending a response to the client. Dedicating a complete disk to the ZooKeeper transaction log directory prevents I/O processing by other applications from overloading the disk.
+    
+You should be sure ZooKeeper directories are configured to use fast disks that can complete synchronization (`fsync`) operations successfully in a timely fashion.
+
+## Provision ZooKeeper directories for master nodes
+You should put the `/var/lib/dcos/exhibitor/zookeeper/transactions` directory for each master node on a separate disk.
```
[the remark this answers is not in the document]
Comment (lsgunnlsgunn) [reply pages/1.13/installing/production/system-requirements/zookeeper-requirements/index.md]: Removed the exhibitor/zookeeper/transactions reference. Hid the parameters for dataDir and dataLogDir.

File: pages/1.13/installing/production/system-requirements/zookeeper-requirements/index.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+<code>dd if=/dev/zero of=/tmp/test1.img bs=1024 count=1</code><br>
+<code>dd if=/dev/zero of=/tmp/test2.img bs=1024 count=1000</code>
+
+## Isolate master nodes from agent nodes
+
+Isolate master nodes from agent nodes to prevent memory and CPU contention, especially when workloads are started on the agents.
+
+The most common ZooKeeper issues occur when ZooKeeper transaction log files are stored on the same solid state drive (SSD) as other components where ZooKeeper write performance becomes a bottleneck that slows down or delays processing of other worker threads.
+    
+ZooKeeper should not share resources with any other processes or services. ZooKeeper writes new transactions to the log before performing updates or sending a response to the client. Dedicating a complete disk to the ZooKeeper transaction log directory prevents I/O processing by other applications from overloading the disk.
```
[the remark this answers is not in the document]
Comment (lsgunnlsgunn) [reply pages/1.13/installing/production/system-requirements/zookeeper-requirements/index.md]: Revised slightly, but need the security team to look at it.

File: pages/1.13/installing/production/system-requirements/zookeeper-requirements/index.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+# Why planning ZooKeeper resources is important
+As a fundamental component of the DC/OS platform architecture, ZooKeeper performs several critical tasks for the DC/OS cluster. For example, ZooKeeper identifies which master node is used as the leader and coordinates the leader selection so that this information is available to the other master nodes, agents, and schedulers. ZooKeeper also maintains state information for Marathon-orchestrated services and persists this information in its transaction logs and snapshots.
+
+Because ZooKeeper status, operation, and performance can directly affect the stability, resiliency, and performance of the DC/OS cluster, it is important to optimize your ZooKeeper configuration to handle the intended cluster workload effectively and efficiently. For example, issues with ZooKeeper write performance often lead to latency-related problems and degraded cluster performance.
+
+# Transaction logs and snapshots
```
[the remark this answers is not in the document]
Comment (lsgunnlsgunn) [reply pages/1.13/installing/production/system-requirements/zookeeper-requirements/index.md]: Removed references to the location parameters.

File: pages/1.13/installing/production/system-requirements/zookeeper-requirements/index.md [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+layout: layout.pug
+title: ZooKeeper resources
+navigationTitle: ZooKeeper resources
+menuWeight: 10
+excerpt: Requirements and recommendations for ZooKeeper in a DC/OS cluster
+---
+ZooKeeper is a centralized coordination service that stores, maintains, and synchronizes information for distributed systems. ZooKeeper and its management service maintain state information and record details of node activity in a data directory. As changes are made to the cluster, those changes are recorded in the ZooKeeper transaction log. When the transaction log grows too large, ZooKeeper creates a snapshot of the current state of cluster nodes.
+
+# Why planning ZooKeeper resources is important
+As a fundamental component of the DC/OS platform architecture, ZooKeeper performs several critical tasks for the DC/OS cluster. For example, ZooKeeper identifies which master node is used as the leader and coordinates the leader selection so that this information is available to the other master nodes, agents, and schedulers. ZooKeeper also maintains state information for Marathon-orchestrated services and persists this information in its transaction logs and snapshots.
```
Comment (jongiddy) [inline pages/1.13/installing/production/system-requirements/zookeeper-requirements/index.md]: "and persists this information in its transaction logs and snapshots" seems out of place here. This describes how ZooKeeper persists data, but it is in no way specific to Marathon state. Can we just remove it, or make it an introductory sentence for the "Transaction logs and snapshots" subsection? e.g. "ZooKeeper persists data in its transaction logs and snapshots."

File: pages/1.13/installing/production/system-requirements/zookeeper-requirements/index.md [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+Because ZooKeeper status, operation, and performance can directly affect the stability, resiliency, and performance of the DC/OS cluster, it is important to optimize your ZooKeeper configuration to handle the intended cluster workload effectively and efficiently. For example, issues with ZooKeeper write performance often lead to latency-related problems and degraded cluster performance.
+
+# Transaction logs and snapshots
+In planning your cluster deployment, there are two important points to keep in mind for ZooKeeper:
+- ZooKeeper processes each read or write operation as a complete transaction rather than performing partial or incremental updates when there are changes to nodes in the cluster.
+- ZooKeeper does not remove old snapshots and log files from the file system. You must manage these files manually.
+
+For a DC/OS cluster, both the snapshots and transaction log are stored in the `/var/lib/dcos/exhibitor` directory. Using a dedicated log device for ZDC/OS exhibitor (`dcos-exhibitor`) log files helps to avoid resource contention and latency issues.
```
Comment (jongiddy) [inline pages/1.13/installing/production/system-requirements/zookeeper-requirements/index.md]: ZDC/OS? Should this just be DC/OS?

File: pages/1.13/installing/production/system-requirements/zookeeper-requirements/index.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (jongiddy) [reply pages/1.13/installing/production/system-requirements/zookeeper-requirements/index.md]: I agree that the first paragraph here is not necessary. The first point is too low-level to be relevant for customers, and as stated, we handle the second point using Exhibitor.

File: pages/1.13/installing/production/system-requirements/zookeeper-requirements/index.md [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+## Isolate master nodes from agent nodes
+
+Isolate master nodes from agent nodes to prevent memory and CPU contention, especially when workloads are started on the agents.
+
+The most common ZooKeeper issues occur when ZooKeeper transaction log files are stored on the same solid state drive (SSD) as other components where ZooKeeper write performance becomes a bottleneck that slows down or delays processing of other worker threads.
+    
+ZooKeeper should not share resources with any other processes or services. ZooKeeper writes new transactions to the log before performing updates or sending a response to the client. Dedicating disk space for the ZooKeeper transaction log directory prevents I/O processing by other applications from overloading the disk.
+    
+You should be sure ZooKeeper directories are configured to use fast disks that can complete synchronization (`fsync`) operations successfully in a timely fashion.
```
Comment (jongiddy) [inline pages/1.13/installing/production/system-requirements/zookeeper-requirements/index.md]: I suggest moving the potential problems text here as:

If there are disk latency issues writing to the transaction log, ZooKeeper may log messages similar to the following:

`WARN SyncThread:14  fsync-ing the write ahead log in SyncThread:14 took 14818ms which will adversely effect operation latency. See the ZooKeeper troubleshooting guide`

(Note I have removed network latency - this is only relevant if the disk if network mounted - that is, it is a specific way in which a disk might have high latency).

File: pages/1.13/installing/production/system-requirements/zookeeper-requirements/index.md [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+Because ZooKeeper status, operation, and performance can directly affect the stability, resiliency, and performance of the DC/OS cluster, it is important to optimize your ZooKeeper configuration to handle the intended cluster workload effectively and efficiently. For example, issues with ZooKeeper write performance often lead to latency-related problems and degraded cluster performance.
+
+# Transaction logs and snapshots
+In planning your cluster deployment, there are two important points to keep in mind for ZooKeeper:
+- ZooKeeper processes each read or write operation as a complete transaction rather than performing partial or incremental updates when there are changes to nodes in the cluster.
+- ZooKeeper does not remove old snapshots and log files from the file system. You must manage these files manually.
+
+For a DC/OS cluster, both the snapshots and transaction log are stored in the `/var/lib/dcos/exhibitor` directory. Using a dedicated log device for ZDC/OS exhibitor (`dcos-exhibitor`) log files helps to avoid resource contention and latency issues.
+
+# Identifying potential problems
```
Comment (jongiddy) [inline pages/1.13/installing/production/system-requirements/zookeeper-requirements/index.md]: This is one very specific problem. I suggest removing this and adding some text below (see other comment).

File: pages/1.13/installing/production/system-requirements/zookeeper-requirements/index.md [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+Isolate master nodes from agent nodes to prevent memory and CPU contention, especially when workloads are started on the agents.
+
+The most common ZooKeeper issues occur when ZooKeeper transaction log files are stored on the same solid state drive (SSD) as other components where ZooKeeper write performance becomes a bottleneck that slows down or delays processing of other worker threads.
+    
+ZooKeeper should not share resources with any other processes or services. ZooKeeper writes new transactions to the log before performing updates or sending a response to the client. Dedicating disk space for the ZooKeeper transaction log directory prevents I/O processing by other applications from overloading the disk.
+    
+You should be sure ZooKeeper directories are configured to use fast disks that can complete synchronization (`fsync`) operations successfully in a timely fashion.
+
+## Provision ZooKeeper directories for master nodes
```
Comment (jongiddy) [inline pages/1.13/installing/production/system-requirements/zookeeper-requirements/index.md]: I'm not sure this captures the ideas completely correctly.

The `/var/lib/dcos` directory must be on a fast local disk. This is required to ensure that files from several DC/OS components have low latency. However, as the disk is still shared between multiple services, this configuration will reduce contention, but will not prevent seeking.

ZooKeeper can benefit from specifically isolating the ZooKeeper transaction log directory (`/var/lib/dcos/exhibitor/zookeeper/transactions`) on a separate device, separate even from `/var/lib/dcos/exhibitor/zookeeper/snapshot`. In this configuration, we would expect to benefit from the lack of seeking. The difficulty with this advice is that there is a limit on how many fast local disks we can expect a node to provide.

Hence, I suggest reducing this text to something like:

You should put the `/var/lib/dcos` directory for each master node on a separate dedicated disk to reduce latency issues. Using a dedicated partition may not be sufficient. Putting the log on a busy or shared device will adversely affect performance.

File: pages/1.13/installing/production/system-requirements/zookeeper-requirements/index.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (lsgunnlsgunn) [reply pages/1.13/installing/production/system-requirements/zookeeper-requirements/index.md]: I think both comments refer to the Transaction logs and snapshots section. Modified to just cover the general point about where the files live and the use of exhibitor.

File: pages/1.13/installing/production/system-requirements/zookeeper-requirements/index.md (hunk 5 above)
Comment (lsgunnlsgunn) [reply pages/1.13/installing/production/system-requirements/zookeeper-requirements/index.md]: Moved under the transaction log /snapshot heading.

File: pages/1.13/installing/production/system-requirements/zookeeper-requirements/index.md (hunk 6 above)
Comment (lsgunnlsgunn) [reply pages/1.13/installing/production/system-requirements/zookeeper-requirements/index.md]: oops

File: pages/1.13/installing/production/system-requirements/zookeeper-requirements/index.md (hunk 7 above)
Comment (lsgunnlsgunn) [reply pages/1.13/installing/production/system-requirements/zookeeper-requirements/index.md]: changed.

**Pull request merged**